File: DCP-1816/solution.kt
class TimeMap<K, V> {
    private val storage = mutableMapOf<K, MutableList<Pair<Int, V>>>()
    
    fun set(key: K, value: V, timestamp: Int) {
        storage.getOrPut(key) { mutableListOf() }
            .add(Pair(timestamp, value))
    }
    
    fun get(key: K, timestamp: Int): V? {
        val values = storage[key] ?: return null
        
        var left = 0
        var right = values.size - 1
        var result: V? = null
        
        while (left <= right) {
            val mid = (left + right) / 2
            val currentTimestamp = values[mid].first
            
            if (currentTimestamp <= timestamp) {
                result = values[mid].second
                left = mid + 1
            } else {
                right = mid - 1
            }
        }
        
        return result
    }
}

File: DCP-1816/solution.swift
// DCP-1816
// Implement custom map with timeframe function

import Foundation

class TimeMap<K: Hashable, V> {
    private var storage: [K: [(timestamp: Int, value: V)]] = [:]
    
    // O(1) operation, since we are just appending the value to the storage.
    func set(key: K, value: V, timestamp: Int) {
        if storage[key] == nil {
            storage[key] = []
        }
        storage[key]?.append((timestamp: timestamp, value: value))
    }
    // O(log n) operation, since we are using binary search to find the value.
    func get(key: K, timestamp: Int) -> V? {
        guard let values = storage[key] else { return nil }
        
        var left = 0
        var right = values.count - 1
        var result: V? = nil
        
        while left <= right {
            let mid = (left + right) / 2
            let currentTimestamp = values[mid].timestamp
            
            if currentTimestamp <= timestamp {
                result = values[mid].value
                left = mid + 1
            } else {
                right = mid - 1
            }
        }
        
        return result
    }
}